src/Hacknet/HacknetHelpers.tsx

```diff
@@ -460,139 +463,137 @@ export function updateHashManagerCapacity(): void {
   Player.hashManager.updateCapacity(total);
 }
 
-export function purchaseHashUpgrade(upgName: string, upgTarget: string, count = 1): boolean {
+export function purchaseHashUpgrade(upgName: HashUpgradeEnum, upgTarget: string, count = 1): Result {
   if (!(Player.hashManager instanceof HashManager)) {
-    console.error(`Player does not have a HashManager`);
-    return false;
+    exceptionAlert(new Error("Player does not have a HashManager"));
+    return { success: false, message: "Player does not have a HashManager" };
   }
 
   // HashManager handles the transaction. This just needs to actually implement
   // the effects of the upgrade
-  if (Player.hashManager.upgrade(upgName, count)) {
-    const upg = HashUpgrades[upgName];
+  const upgradeResult = Player.hashManager.upgrade(upgName, count);
+  if (!upgradeResult.success) {
+    return upgradeResult;
+  }
+
+  const upg = HashUpgrades[upgName];
 
-    switch (upgName) {
-      case "Sell for Money": {
-        Player.gainMoney(upg.value * count, "hacknet");
-        break;
+  switch (upgName) {
+    case HashUpgradeEnum.SellForMoney: {
+      Player.gainMoney(upg.value * count, "hacknet");
+      break;
+    }
+    case HashUpgradeEnum.SellForCorporationFunds: {
+      const corp = Player.corporation;
+      if (corp === null) {
+        Player.hashManager.refundUpgrade(upgName, count);
+        return { success: false, message: "You have not created a corporation." };
       }
-      case "Sell for Corporation Funds": {
-        const corp = Player.corporation;
-        if (corp === null) {
-          Player.hashManager.refundUpgrade(upgName, count);
-          return false;
+      corp.gainFunds(upg.value * count, "hacknet");
+      break;
+    }
+    case HashUpgradeEnum.ReduceMinimumSecurity: {
+      try {
```

Comment:
The use of try is awkward and a bit dangerous here. Do we need it? If one of the functions we're calling uses throw for "standard" error reporting, can we switch it to result?